nomos-services/data-availability/sampling/src/lib.rs

```diff
@@ -362,17 +439,57 @@ where
         loop {
             tokio::select! {
                 Some(service_message) = service_resources_handle.inbound_relay.recv() => {
-                    Self::handle_service_message(service_message, &mut network_adapter,  &storage_adapter,  &mut sampler).await;
+                    Self::handle_service_message(
+                        service_message,
+                        &mut network_adapter,
+                        &storage_adapter,
+                        &mut sampler,
+                        commitments_wait_duration,
+                    ).await;
                 }
                 Some(sampling_message) = sampling_message_stream.next() => {
-                    Self::handle_sampling_message(sampling_message, &mut sampler, &storage_adapter, &share_verifier).await;
+                    Self::handle_sampling_message(
+                        sampling_message,
+                        &mut sampler,
+                        &storage_adapter,
+                        &share_verifier
+                    ).await;
+                }
+                Some(commitments_message) = commitments_message_stream.next() => {
+                    Self::handle_commitments_message(
+                        &storage_adapter,
+                        commitments_message
+                    ).await;
                 }
                 // cleanup not on time samples
                 _ = next_prune_tick.tick() => {
                     sampler.prune();
                 }
-
             }
         }
     }
 }
+
+async fn wait_commitments(
+    mut stream: Pin<Box<dyn Stream<Item = CommitmentsEvent> + Send>>,
+    wait_duration: Duration,
+    sender: oneshot::Sender<Option<DaSharesCommitments>>,
+    requested_blob_id: BlobId,
+) {
+    let _ = tokio::time::timeout(wait_duration, async {
+        while let Some(message) = stream.next().await {
+            if let CommitmentsEvent::CommitmentsSuccess {
+                blob_id,
+                commitments,
+            } = message
+            {
+                if blob_id == requested_blob_id {
+                    let _ = sender.send(Some(*commitments));
+                    return;
+                }
+            }
+        }
+        let _ = sender.send(None);
+    })
+    .await;
+}
```

Comment:
Couldn't this be part of the tokio stream? We should avoid spawning tasks if possible.

Reply:
Let me take a shot a it

Reply:
Yeah, I tried using futuresunordered, but it complains about some bounds. We can go ahead with that for now. But it need to be a better solution rather than spawning this tasks.

Reply:
Lets open an issue for this please @bacv

Reply:
[Issue](https://github.com/logos-co/nomos/issues/1603)